tests/unit/neptune/new/internal/utils/test_git.py

```diff
@@ -59,3 +61,84 @@ def test_getting_git_info(self, repo_mock):
             branch="master",
             remotes=[],
         )
+
+
+class TestDiffTracker:
+    def test_get_head_index_diff(self):
+        repo_mock = MagicMock()
+        repo_mock.git.diff.return_value = "some_diff"
+        repo_mock.head.name = "HEAD"
+
+        tracker = DiffTracker(repo_mock)
+
+        diff = tracker.get_head_index_diff()
+
+        repo_mock.git.diff.assert_called_once_with("HEAD")
+        assert diff == "some_diff"
+
+    def test_upstream_index_diff_tracking_branch_present(self):
+        repo_mock = MagicMock()
+        repo_mock.git.diff.return_value = "some_diff"
+        tracking_branch = MagicMock()
+        tracking_branch.commit.hexsha = "sha1234"
+        repo_mock.active_branch.tracking_branch.return_value = tracking_branch
+
+        tracker = DiffTracker(repo_mock)
+
+        diff = tracker.get_upstream_index_diff()
+
+        repo_mock.git.diff.assert_called_once_with("sha1234")
+        assert diff == "some_diff"
+        assert tracker.upstream_commit_sha == "sha1234"
+
+    def test_upstream_index_diff_tracking_branch_not_present(self):
+        repo_mock = MagicMock()
+        repo_mock.git.diff.return_value = "some_diff"
+        repo_mock.active_branch.tracking_branch.return_value = None
+
+        tracking_branch = MagicMock()
+        branch = MagicMock()
+        branch.tracking_branch.return_value = tracking_branch
+        repo_mock.branches = [branch, branch, branch]
+
+        repo_mock.is_ancestor.return_value = True
+        ancestor = MagicMock()
+        ancestor.hexsha = "sha1234"
+        repo_mock.merge_base.return_value = [ancestor, ancestor]
+
+        tracker = DiffTracker(repo_mock)
+
+        diff = tracker.get_upstream_index_diff()
+
+        assert diff == "some_diff"
+        assert tracker.upstream_commit_sha == "sha1234"
+
+        repo_mock.git.diff.assert_called_once_with("sha1234")
+        repo_mock.active_branch.tracking_branch.assert_called_once()
+
+        assert repo_mock.merge_base.call_count == 3
+        assert repo_mock.is_ancestor.call_count == 5  # 6 ancestors - 1 case when most_recent_ancestor was None
+
+    def test_detached_head(self):
+        repo_mock = MagicMock()
+        repo_mock.active_branch.tracking_branch.side_effect = TypeError
+        repo_mock.git.diff = MagicMock()
+
+        tracker = DiffTracker(repo_mock)
+        diff = tracker.get_upstream_index_diff()
+
+        assert diff is None
+        assert tracker.upstream_commit_sha is None
+        repo_mock.git.diff.assert_not_called()
+
+    @patch("neptune.internal.utils.git.File")
+    @patch("neptune.metadata_containers.Run")
+    def test_track_uncommitted_changes(self, mock_run, mock_file):
+        mock_tracker = MagicMock()
+
+        with patch("neptune.internal.utils.git.DiffTracker.from_git_ref", return_value=mock_tracker):
```

Comment:
Check lines 36 and 37 in this file. This and similar mocks could be replaced with a simplified syntax (I think so).

Reply:
I think that this way I won't be able to access the mock_tracker object to check the args with which it's methods were called.

Reply:
You're probably right :wink:. We should be able to simplify it but I'm struggling figuring it out.